machinelearning/caffe/README.md

```diff
@@ -0,0 +1,220 @@
+# Caffe
+
+Caffe is a deep learning framework that enables the possibility to run models in GPU/CPU, it enhances the modularity and speed of models.
+
+#### What you need:
+
+To build this container, you will need the following:
+
+ - Singularity, which you can download and install from [here](https://github.com/sylabs/singularity).
+ - A text editor, like: `micro`, `vim` or `nano`.
+ - Root access.
+
+#### Set up:
+
+Create a folder where all of your files to the caffe singularity container including the def file will reside. In this example, we will call the folder `caffehost` and inside we will create two (2) additional folders to run the example: `data` and `examples`.
+
+Your `caffehost` folder should look like this:
+
+```
+caffehost/
+|-- data
+|-- examples
+|-- build
+|   |--examples
+`-- caffe.def
+```
+Where `caffe.def` is the definition file we will use to build the Singularity container.
+
+#### Getting started: Building the Singularity Caffe container:
+
+To build your container, move into the `caffehost` folder and run:
+
+```
+sudo singularity build caffe.sif caffe.def
+```
+
+We will go through a detailed explanation of the sections inside this definition file `caffe.def` :
+
+On the `%help` section you will find a description of the container, which versions of every software is supported and over which OS image it runs.
+
+While on the `%environment` section we will define the environment variables needed for the installation. This will be needed for running the container in an example explained later on.
+
+On the `%post` section we describe all the needed dependencies, libraries and source code to be able to build the caffe container. For example, some modifications include specifying that this installation is only **CPU based**.
+
+#### Binding folders from host to container
+
+Before running the container, we will need to bind some folders from host to container in order to follow the example.
+
+Run the following:
+
+```
+$ export SINGULARITY_BINDPATH="/caffehost/data:/caffe/data,/caffehost/examples:/caffe/examples, /caffehost/build/examples:/caffe/build/examples"
+```
+
+This will bind the folder on your host `/caffehost/data` to `/caffe/data` in the container and `/caffehost/examples` to `/caffe/examples` respectively.
+
+After this needed binding we can shell into the container and run:
+
+```
+$ sudo singularity shell caffe.sif
+```
+
+#### Test your installation by running an example:
+
+We can start executing the example to test our installation of Caffe inside the container:
+
+This example is based on the `MNIST` example that appears on the Caffe website found [here](http://caffe.berkeleyvision.org/gathered/examples/mnist.html)
+
+Move into the `caffehost` folder and run:
+
+```
+cd data
+cp -a /caffe/data/. .
```

Comment:
Here you're bind mounting the `caffehost/data` directory to `/caffe/data/` ... So, you're overlaying the host `data` over what exists in the container. Same as the other directories.

You'll need to initially mount the `caffehost` directories under `/mnt` or some similar directory, then copy the `/caffe/[blah]` directories to that location. Exit out of the container, and then rerun with the correct bind mounts.